Sources/SourceKitLSP/Swift/SemanticTokens.swift

```diff
@@ -0,0 +1,352 @@
+    Position(line: start.line, utf16index: start.utf16index + length)
+  }
+  public var range: Range<Position> {
+    start..<end
+  }
+
+  public init(
+    name: String? = nil,
+    start: Position,
+    length: Int,
+    kind: Kind,
+    modifiers: Modifiers = []
+  ) {
+    self.start = start
+    self.length = length
+    self.kind = kind
+    self.modifiers = modifiers
+  }
+
+  /// The token type. Represented using an int to make the conversion to
+  /// LSP tokens efficient.
+  public enum Kind: UInt32, CaseIterable, Hashable {
+    case comment = 0
+    case keyword
+    case modifier
+    case regexp
+    case `operator`
+    case namespace
+    case type
+    case `struct`
+    case `class`
+    case interface
+    case `enum`
+    case typeParameter
+    case function
+    case macro
+    case method
+    case variable
+    case parameter
+    case property
+    case label
+    case number
+    case string
+
+    /// The name of the token type used by LSP.
+    var lspName: String {
+      switch self {
+      case .comment: return "comment"
+      case .keyword: return "keyword"
+      case .modifier: return "modifier"
+      case .regexp: return "regexp"
+      case .operator: return "operator"
+      case .namespace: return "namespace"
+      case .type: return "type"
+      case .struct: return "struct"
+      case .class: return "class"
+      case .interface: return "interface"
+      case .enum: return "enum"
+      case .typeParameter: return "typeParameter"
+      case .function: return "function"
+      case .macro: return "macro"
+      case .method: return "method"
+      case .variable: return "variable"
+      case .parameter: return "parameter"
+      case .property: return "property"
+      case .label: return "label"
+      case .number: return "number"
+      case .string: return "string"
+      }
+    }
+  }
+
+  /// Additional metadata about a token.
+  public struct Modifiers: OptionSet, CaseIterable, Hashable {
+    public static let declaration = Self(rawValue: 1 << 0)
+    public static let definition = Self(rawValue: 1 << 1)
+    public static let readonly = Self(rawValue: 1 << 2)
+    public static let `static` = Self(rawValue: 1 << 3)
+    public static let deprecated = Self(rawValue: 1 << 4)
+    public static let abstract = Self(rawValue: 1 << 5)
+    public static let async = Self(rawValue: 1 << 6)
+    public static let modification = Self(rawValue: 1 << 7)
+    public static let documentation = Self(rawValue: 1 << 8)
+    public static let defaultLibrary = Self(rawValue: 1 << 9)
+
+    /// All available modifiers, in order
+    public static let allCases: [Self] = [
+      .declaration,
+      .definition,
+      .readonly,
+      .static,
+      .deprecated,
+      .abstract,
+      .async,
+      .modification,
+      .documentation,
+      .defaultLibrary,
+    ]
+
+    public let rawValue: UInt32
+
+    /// The name of the modifier used by LSP.
+    public var lspName: String? {
```

Comment:
I don't think we can do so, since `Modifier` is an `OptionSet`, where the `rawValue` potentially is a bitwise-ORed combination of multiple modifiers listed in `allCases`.

Reply:
Also we have to be careful not to have cyclic definitions here (i.e. `allCases` references the cases which references `init(rawValue:)` which references `allCases`).

I would argue that the current solution would avoid introducing a lot of complexity for an assertion that we are unlikely to accidentally run into anyway (since these modifiers are rarely constructed from raw values outside of the static constants).

Reply:
Good point. I didn’t consider that it was an option set. Leaving it as-is sounds good to me.